File: leetcode/230-kth-smallest-element-in-a-bst.js
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/ (top-interview-questions)

// Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
  let visited = 0;

  let stack = [];
  let node = root;
  while (stack.length > 0 || node) {
    while (node) {
      stack.push(node);
      node = node.left;
    }
    node = stack.pop();
    visited += 1;
    if (visited === k) {
      return node.val;
    }
    node = node.right;
  }

};

File: leetcode/230-kth-smallest-element-in-a-bst.py
# https://leetcode.com/problems/kth-smallest-element-in-a-bst/ (top-interview-questions)

from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = list()
        node = root
        counter = 0 # number of processed nodes
        while stack or node:
            while node:
                stack.append(node)
                node = node.left
            node = stack.pop()
            counter += 1
            if counter == k:
                return node.val
            node = node.right
